### PR TITLE
apps/testcase/wifi_manager : add ITC testcase for WiFi manager

### PR DESCRIPTION
**Pull request by manoj11588612**
New testcase's :
1.itc_wifi_manager_connect_ap_config_p  -> use to connect a AP with all SSID configuration.
2.itc_wifi_manager_save_get_config_p    -> use to save the configuration parameter and then geting saved configuration parameter
3.itc_wifi_manager_get_remove_config_p  -> use to get configuration parameter and removing ap configuration after get.
4.itc_wifi_manager_remove_get_config_p  -> use to remove then get configuration parameter of ap.
5.itc_wifi_manager_get_config_n         -> This is a negative tc in which we try to get ap configuration parameter even without saving.
6.itc_wifi_manager_connect_ap_config_n  -> This is a negative tc in which we try to connect with ap with invalid parameter.

Signed-off-by: manoj <manoj.g2@samsung.com>

### CODE REVIEW

**Comment by manoj11588612** [comment]
@sunghan-chang @pillip8282  @samegautam @sangwon03 
Please review

**Comment by seinfra** [comment]
Target : [ff6f2e6f91f482a2003dba81711a754b7ba08624] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 1]
```diff
@@ -288,6 +288,241 @@ static void itc_wifi_manager_connect_disconnect_ap_p(void)
...
+	strncpy(config.passphrase, TEST_PASSWORD, config.passphrase_length + 1);
+	config.ap_auth_type = (wifi_manager_ap_auth_type_e)TEST_AUTH_TYPE;
+	config.ap_crypto_type = (wifi_manager_ap_crypto_type_e)TEST_CRYPTO_TYPE;
+	printf("AP config: %s(%d), %s(%d), %d %d\n", config.ssid, config.ssid_length, config.passphrase, \
+			config.passphrase_length, config.ap_auth_type, config.ap_crypto_type);
+
+	ret =  wifi_manager_connect_ap_config(&config, &reconfig);
+	TC_ASSERT_EQ_CLEANUP("wifi_manager_connect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+	WIFITEST_WAIT;
+	sleep(SEC_5);
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: why there is sleep?

Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: for connecting to a ap its required some time. for that we use sleep.
in other previous developed TC also we used sleep.

**Comment by seinfra** [comment]
Target : [fc94b1e59416f931947c695f4ac3b34d9766de00] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fc94b1e59416f931947c695f4ac3b34d9766de00] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fc94b1e59416f931947c695f4ac3b34d9766de00] -

**Comment by seinfra** [comment]
Target : [4415b5715250a2da3f3fc4fbf9d8cc4d44a0781b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4415b5715250a2da3f3fc4fbf9d8cc4d44a0781b] - Code Rule Check OK.

**Comment by samegautam** [comment]
@sunghan-chang @pillip8282 Please review the PR

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 2]
```diff
@@ -288,6 +288,240 @@ static void itc_wifi_manager_connect_disconnect_ap_p(void)
...
+	config.ap_crypto_type = (wifi_manager_ap_crypto_type_e)TEST_CRYPTO_TYPE;
+	printf("AP config: %s(%d), %s(%d), %d %d\n", config.ssid, config.ssid_length, config.passphrase, \
+			config.passphrase_length, config.ap_auth_type, config.ap_crypto_type);
+
+	ret =  wifi_manager_connect_ap_config(&config, &reconfig);
+	TC_ASSERT_EQ_CLEANUP("wifi_manager_connect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+	WIFITEST_WAIT;
+
+	ret = wifi_manager_deinit();
+	TC_ASSERT_EQ("wifi_manager_deinit", ret, WIFI_MANAGER_SUCCESS);
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: if wifi manager succeeded to connect AP, then it have to wait the disconnect event when it called wifi_manager_deinit().
please check it.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 3]
```diff
@@ -288,6 +288,240 @@ static void itc_wifi_manager_connect_disconnect_ap_p(void)
...
+	strncpy(config.ssid, TEST_SSID, config.ssid_length + 1);
+	strncpy(config.passphrase, TEST_PASSWORD, config.passphrase_length + 1);
+	config.ap_auth_type = (wifi_manager_ap_auth_type_e)TEST_AUTH_TYPE;
+	config.ap_crypto_type = (wifi_manager_ap_crypto_type_e)TEST_CRYPTO_TYPE;
+	printf("AP config: %s(%d), %s(%d), %d %d\n", config.ssid, config.ssid_length, config.passphrase, \
+			config.passphrase_length, config.ap_auth_type, config.ap_crypto_type);
+
+	ret =  wifi_manager_connect_ap_config(&config, NULL);
+	TC_ASSERT_NEQ_CLEANUP("wifi_manager_connect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: same as above

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 4]
```diff
@@ -288,6 +288,240 @@ static void itc_wifi_manager_connect_disconnect_ap_p(void)
...
+	ret =  wifi_manager_save_config(&config);
+	TC_ASSERT_EQ_CLEANUP("wifi_manager_save_config", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+
+	ret =  wifi_manager_get_config(&gconfig);
+	TC_ASSERT_EQ_CLEANUP("wifi_manager_get_config", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+
+	ret =  wifi_manager_remove_config();
+	TC_ASSERT_EQ_CLEANUP("wifi_manager_remove_config", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
+
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: I think you should call wifi_manager_get_config() again to check that wifi manager removes the configration well.

Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: I alraedy created one more Testcase for this scenario. 
Please check  testcase "itc_wifi_manager_remove_get_config_p"

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c (hunk 2 above)
Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: code changes and commit

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c (hunk 3 above)
Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: code changes and commit

**Comment by SeonghoByeon** [comment]
I have verifed that all 22 itc functions including additional features work successfully on artik053 board.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c (hunk 4 above)
Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Both Testcase are fordifferent purpose.
Like First TC is just for positive scenario likeideally it shopuld work but in second Testcase we arev also checking for scenario in whiich after removing again i called get config which is a negative scenario and use for checking invalid parameter issue also.
please suggest if you are not agree on my above comment.

**Comment by manoj11588612** [comment]
@pillip8282 @SeonghoByeon @samegautam @sunghan-chang @sangwon03 
I removed all sleep() call
Please review

**Comment by seinfra** [comment]
Target : [23a6f2ca810885cd468b1eed80d97470cf48676a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [23a6f2ca810885cd468b1eed80d97470cf48676a] - Code Rule Check Result:
apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c:73: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [a93105b74c1b843c86ce60649383f3c802522fad] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [a93105b74c1b843c86ce60649383f3c802522fad] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 5]
```diff
@@ -805,10 +1078,9 @@ int wifi_manager_itc(int argc, FAR char *argv[])
 	itc_wifi_manager_set_mode_p();
 
 	itc_wifi_manager_get_mode_p(); // set softap mode inside this function
-
 	itc_wifi_manager_connect_disconnect_ap_p(); // change to station mode and try to connect to ap
 
-	itc_wifi_manager_scan_ap_p(); // Reinitialized wifi manager with the callback hander for scan results
+	//itc_wifi_manager_scan_ap_p(); // Reinitialized wifi manager with the callback hander for scan results
```
Comment (SeonghoByeon) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Hi @manoj11588612,
The reason for not doing the scanning here is probably the state error in the WiFi Manager.
Please add WIFITEST_WAIT right after calling wifi_manager_scan_ap(),
and then, add WIFITEST_SIGNAL in wifi_scan_ap_done_cb() callback function.
(WIFITEST_SIGNAL should be called within wifi_scan_ap_done_cb() whether it succeeded or not.)
Then, the scanning operation does not need to be done last in wifi_manager_itc(),
and the reconnect operation will also be successful.

Comment (manoj11588612) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: I agree with your comment.
I added your comment and checked . ALL testcase gettign passed.
Please review.

**Comment by seinfra** [comment]
Target : [418af845f9a494e840b79b997ff070f7592c685c] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 6]
```diff
@@ -64,19 +64,19 @@ static wifi_manager_cb_s wifi_callbacks = {
...
 
-static void wifi_sta_disconnected_cb(void)
+void wifi_sta_disconnected_cb(void)
 {
 	printf("wifi_sta_disconnected: send signal!!! \n");
 	WIFITEST_SIGNAL;
 }
 
-static void wifi_scan_ap_done_cb(wifi_manager_scan_info_s **scan_info, wifi_manager_scan_result_e res)
+void wifi_scan_ap_done_cb(wifi_manager_scan_info_s **scan_info, wifi_manager_scan_result_e res)
```
Comment (SeonghoByeon) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Hi @manoj11588612,
I have confirmed that all of the issues mentioned above have been resolved.
However, in order to cope with scanning failure that may occur in the (cypress) wifi device driver,
please add WIFITEST_SIGNAL before returning in case of WIFI_SCAN_FAIL.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 7]
```diff
@@ -389,11 +676,9 @@ static void itc_wifi_manager_reconnect_p(void)
 	TC_ASSERT_EQ_CLEANUP("wifi_manager_connect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
 	WIFITEST_WAIT;
 
-	sleep(SEC_5);
 	ret = wifi_manager_connect_ap(&config);
 	TC_ASSERT_EQ_CLEANUP("wifi_manager_connect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_disconnect_ap(); wifi_manager_deinit());
```
Comment (SeonghoByeon) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: This assert should be changed to TC_ASSERT_NEQ_CLEANUP, because a connection trial in the connected state returns WIFI_MANAGER_FAIL.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 8]
```diff
@@ -389,11 +676,9 @@ static void itc_wifi_manager_reconnect_p(void)
...
 
-	sleep(SEC_5);
 	ret = wifi_manager_disconnect_ap();
 	TC_ASSERT_EQ_CLEANUP("wifi_manager_disconnect_ap", ret, WIFI_MANAGER_SUCCESS, wifi_manager_deinit());
```
Comment (SeonghoByeon) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Please add WIFITEST_WAIT here, to prevent a state error.

Comment (SeonghoByeon) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Then, you can call itc_wifi_manager_reconnect_p() at the end of the main function.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 9]
```diff
@@ -805,33 +1082,47 @@ int wifi_manager_itc(int argc, FAR char *argv[])
...
 
 	itc_wifi_manager_connect_long_ssid_p();
-
 	itc_wifi_manager_average_joining_ap(); // Disable DoS preventation from Access Point
 
 	itc_wifi_manager_average_leaving_ap(); // Disable DoS preventation from Access Point
 
 	itc_wifi_manager_average_stoping_ap(); // Disable DoS preventation from Access Point
 
 	itc_wifi_manager_success_ratio_ap();
```
Comment (SeonghoByeon) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Please check that the success ratio works correctly. The output messages are broken.

**Comment by seinfra** [comment]
Target : [91b2bd84c97eb143436ed7729e455c263493be14] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [91b2bd84c97eb143436ed7729e455c263493be14] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [81274c416632aa5d210582d73a74136fd525d890] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [81274c416632aa5d210582d73a74136fd525d890] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 10]
```diff
@@ -52,9 +52,9 @@ static pthread_cond_t g_wifi_manager_test_cond;
 		pthread_mutex_unlock(&g_wifi_manager_test_mutex);\
 	} while (0)
 
-static void wifi_sta_connected_cb(wifi_manager_result_e res); // in station mode, connected to ap
-static void wifi_sta_disconnected_cb(void); // in station mode, disconnected from ap
-static void wifi_scan_ap_done_cb(wifi_manager_scan_info_s **scan_info, wifi_manager_scan_result_e res); // called when scanning ap is done
+void wifi_sta_connected_cb(wifi_manager_result_e res); // in station mode, connected to ap
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Why did you change it to non-static?

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 11]
```diff
@@ -288,6 +290,292 @@ static void itc_wifi_manager_connect_disconnect_ap_p(void)
...
+
+	ret = wifi_manager_deinit();
+	TC_ASSERT_EQ("wifi_manager_deinit", ret, WIFI_MANAGER_SUCCESS);
+
+	TC_SUCCESS_RESULT();
+}
+
+/**
+ * @testcase         itc_wifi_manager_get_remove_config_p
+ * @brief            check initialize, connect,, save ap config, get ap config, remove ap config, de-initalization operations of wifi manager
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: there are no connect API on the function. please check the comment on ITC testcases

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c (hunk 4 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: itc_wifi_manager_remove_get_config_p(void) like below
1) save ap config 2) get ap config 3) remove ap config 4) get ap config
first step 3 are exactly same to itc_wifi_manager_get_remove_config_p() with same input parameter.  So I asked you that itc_wifi_manager_remove_get_config_p() could cover it.

File: apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c [hunk 12]
```diff
@@ -805,34 +1084,48 @@ int wifi_manager_itc(int argc, FAR char *argv[])
 	itc_wifi_manager_set_mode_p();
 
 	itc_wifi_manager_get_mode_p(); // set softap mode inside this function
-
 	itc_wifi_manager_connect_disconnect_ap_p(); // change to station mode and try to connect to ap
 
 	itc_wifi_manager_scan_ap_p(); // Reinitialized wifi manager with the callback hander for scan results
 
+	WIFITEST_WAIT;
```
Comment (pillip8282) [inline apps/examples/testcase/ta_tc/wifi_manager/itc/itc_wifi_manager_main.c]: Instead of calling WIFITEST_WAIT here, it would be better to calling it inside itc_wifi_manager_scan_ap_p().
And I think wifi manager should be deinitialized when each function(itc_wifi*) is ended.

**Comment by seinfra** [comment]
Target : [d5bab75f83bf8424a5caedac28d495a66b3cb8bc] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d5bab75f83bf8424a5caedac28d495a66b3cb8bc] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e9fd1ed5bf7a43105934ecf549b4e524b87313e1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e9fd1ed5bf7a43105934ecf549b4e524b87313e1] - Code Rule Check OK.
